### ISSUE TITLE
I/O API Features

### ISSUE THREAD

**Issue by MrPowers** [pola-rs/polars#2257]
I've worked a lot with the filesystem I/O APIs with Dask and Spark and have a good understanding of the features that users typically desire.  I figured we can build out a roadmap here, so we can increase the likelihood that we'll be exposing an elegant API for end users.  

The purpose is to start brainstorming on these issues.  I'm flexible and willing to change my mind on these.

I made [similar suggestions in Dask](https://github.com/dask/dask/issues/8147) BTW.

## Writing APIs

Users will be interested in writing data to a single file and writing data to multiple files in parallel.  Writing 50GB of data to 1,000 files in parallel should be a lot faster than writing to a single file.

It might be best to expose separate `to_csv` and `to_csvs` APIs because some of the options exposed to users will be different.  With `to_csv` we'll expect the user to specify the filename, whereas with `to_csvs` we should auto-generate the filenames (with a UUID, a la Spark).  So `df.to_csvs("some/dir")` may output these files:

```
part-00000-01234567890123456789012345678912.snappy.csv
part-00001-01234567890123456789012345678912.snappy.csv
```

Then we could let users specify target file sizes like this:

```python
df.to_csvs("some/dir", target_size="100mb")
```

### Disk partitioning

The writers could support Hive-style disk partitioning as follows.

```python
df.to_csvs("some/partitioned_lake1", partition_by=["country"])
```

Here's the files that'd be output:

```
partitioned_lake1/
  country=Argentina/
    part-00044-cf737804-90ea-4c37-94f8-9aa016f6953a.c000.snappy.parquet
  country=China/
    part-00059-cf737804-90ea-4c37-94f8-9aa016f6953a.c000.snappy.parquet
  country=Russia/
    part-00002-cf737804-90ea-4c37-94f8-9aa016f6953a.c000.snappy.parquet
```

### Save Modes

I think the Spark save modes are good:

* overwrite
* append
* errorifexists

That'd let us use syntax like this:

```python
df.to_csvs("some/dir", save_mode="append")
```

Not sure what's the best default `save_mode`.

## Reader APIs

We already have good solutions for reading a single file and multiple files 😎 

### Disk filtering

Suppose we run `df.filter(pl.col("fruits").str.equals("Argentina"))` (syntax may be off) on the `partitioned_lake1` disk partitioned lake above.

I'm assuming there is some way we can use the query optimizer and intelligently determine that we're running the query off of a disk partitioned lake, so the user won't need to manually specify to only grab the file with the Argentina data.

### Parquet Column Pruning

Does the query optimizer / Parquet reader take care of this automatically?

### Parquet predicate pushdown filtering

Query engines can skip row groups based on the metadata in Parquet files.  I'm guessing we can add these filters via the query optimizer without making the user specify them manually.  Are these being applied already?

**Comment by ritchie46** [comment]
Thanks for the suggestions. I think there is merit in your proposals. Some would be relatively easy to implement, some quite some work.

## On csv io
For the csv reader I don't think it would be much faster if we read from multiple files, as we already memory map and parallelize over the no of threads in the thread pool.

For the csv writer this probably is faster.

## Discussion
> I'm assuming there is some way we can use the query optimizer and intelligently determine that we're running the query off of a disk partitioned lake, so the user won't need to manually specify to only grab the file with the Argentina data.

This would require quite some work to implement, as we don't know the literal expressions when they are pushed down (we only have the function that executes the filter)

> Parquet Column Pruning
> Does the query optimizer / Parquet reader take care of this automatically?

Yes!

> Parquet predicate pushdown filtering
> Query engines can skip row groups based on the metadata in Parquet files. I'm guessing we can add these filters via the > query optimizer without making the user specify them manually. Are these being applied already?

What do you mean by filtering rows based on the meta-data? Currently we apply the filter at the scan level, but ignore the meta-data.

# Todo

All in all I think it would be nice to have this implemented. The work that needs to be done from easy to hard:

- [ ] write to multiple files (can be done in `polars-io` writers)
- [ ] write partitioned  (can be done in `polars-io` writers)
- [ ] read csv/ ipc from multiples files (in `polars-lazy` with `globbing` similar to already implemented for parquet)
- [ ] use partitions in reading to optimize predicate pushdown. (in `polars-lazy`)

**Comment by MrPowers** [comment]
Your todo list looks good to me!

Let me know if [this blog post](https://coiled.io/blog/parquet-file-column-pruning-predicate-pushdown/) helps clarify Parquet predicate pushdown filtering.  Parquet files store row groups, each with min/max values.  You can skip entire row groups when reading files for certain queries.  That might be what you're referring to with "Currently we apply the filter at the scan level"?  

Sorry for making the huge issue.  Just wanted to get all the feature requests out there so we don't back ourselves into any corners when building the user facing APIs.  [The Dask Parquet readers](https://docs.dask.org/en/latest/generated/dask.dataframe.read_parquet.html) have a lot of options, but I guess Polars will be simpler cause it only supports one Parquet engine and doesn't support the metadata file.

Thanks for the detailed responses, as always!

**Comment by ritchie46** [comment]
> Let me know if this blog post helps clarify Parquet predicate pushdown filtering. Parquet files store row groups, each with min/max values. You can skip entire row groups when reading files for certain queries. That might be what you're referring to with "Currently we apply the filter at the scan level"?

No at the moment we combine all predicates into a function. And that function is pushed down to the scan level, and so we can already filter rows on the batches we scan.

But simple predicates like `(a > 100)` could use the metadata indeed. This is not yet done, but can be utilized indeed.

**Comment by universalmind303** [comment]
for that todo list, should we also add 

- [ ]  add optional `save_mode` parameter  to `polars-io` writers?

**Comment by ritchie46** [comment]
Yeap..

**Comment by ritchie46** [comment]
#2267

**Comment by ritchie46** [comment]
#2268

**Comment by binga** [comment]
Thanks for the incredible project and the tireless efforts by the contributors!

Just wanted to check if the `to_csvs` API or something similar is on the project roadmap. :)
